pkg/querier/worker/worker.go

```diff
@@ -217,61 +223,130 @@ func (w *querierWorker) AddressAdded(address string) {
 	w.mu.Lock()
 	defer w.mu.Unlock()
 
+	address := instance.Address
 	if m := w.managers[address]; m != nil {
 		return
 	}
 
-	level.Info(w.log).Log("msg", "adding connection", "addr", address)
+	level.Info(w.log).Log("msg", "adding connection", "addr", address, "in-use", instance.InUse)
 	conn, err := w.connect(context.Background(), address)
 	if err != nil {
 		level.Error(w.log).Log("msg", "error connecting", "addr", address, "err", err)
 		return
 	}
 
 	w.managers[address] = newProcessorManager(ctx, w.processor, conn, address)
+	w.instances[address] = instance
+
 	// Called with lock.
 	w.resetConcurrency()
 }
 
-func (w *querierWorker) AddressRemoved(address string) {
-	level.Info(w.log).Log("msg", "removing connection", "addr", address)
+func (w *querierWorker) InstanceRemoved(instance servicediscovery.Instance) {
+	address := instance.Address
+
+	level.Info(w.log).Log("msg", "removing connection", "addr", address, "in-use", instance.InUse)
 
 	w.mu.Lock()
 	p := w.managers[address]
 	delete(w.managers, address)
+	delete(w.instances, address)
 	w.mu.Unlock()
 
 	if p != nil {
 		p.stop()
 	}
+
+	// Re-balance the connections between the available query-frontends / query-schedulers.
+	w.mu.Lock()
+	w.resetConcurrency()
+	w.mu.Unlock()
+}
+
+func (w *querierWorker) InstanceChanged(instance servicediscovery.Instance) {
+	level.Info(w.log).Log("msg", "updating connection", "addr", instance.Address, "in-use", instance.InUse)
+
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	// Skip if there's no manager for it.
+	if m := w.managers[instance.Address]; m == nil {
+		return
+	}
```

Comment:
How can this happen? That would be a bug. Also instead of skipping, we can call `InstanceAdded`.

Reply:
It shouldn't happen. Instead of calling `InstanceAdded` I would prefer to log it as an error, because looks like a bug we should fix. Also, I noticed there's a race condition between `querierWorker.stopping()` and the check whether the service is stopping in other functions, which could potentially lead to create new connections after the unlock in `stopping()`. I addressed in https://github.com/grafana/mimir/pull/3005/commits/7e3f8b7636ef8f13bd25eca57796a0a6eadcbf1e.

WDYT?

Reply:
Logging seems fine to me.

I don't think doing `ServiceContext` check inside the lock prevents a race, as `BasicService` implementation will cancel the context before calling `stop` function, and this is in no way related to the locks used inside the worker.

Any extra connection created after context has been cancelled will be closed by `m.stop()` call in `stopping` function of the worker.

Reply:
OK, I see what you mean now. Sounds good.